walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,66 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> unavailable_Time = new ArrayList<TimeRange>();
+    List<TimeRange> available_Time = new ArrayList<TimeRange>();
+
+    //Duration greater than a day is not valid, return empty list
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return available_Time;
+    }
+
+    //If no required attendees or any events going on then it can be held at whatever time
+    if (request.getAttendees().isEmpty() || events.isEmpty()) {
+      available_Time.add(TimeRange.WHOLE_DAY);
+      return available_Time;
+    }
+    
+    //Add unavailable times if required attendees must also be at other events
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (request.getAttendees().contains(attendee)) {
+          unavailable_Time.add(event.getWhen());
+          break;
+        }
+      }
+    }
+
+    //If there are no unvailable times, return the whole day
+    if (unavailable_Time.isEmpty()) {
+      available_Time.add(TimeRange.WHOLE_DAY);
+      return available_Time;
+    }
+
+    //Remove unavailable TimeRanges that are contained in other TimeRanges
+    for (int i = 0; i < unavailable_Time.size() - 1; i++) {
+      for (int j = i + 1; j < unavailable_Time.size(); j++) {
+        if (unavailable_Time.get(i).contains(unavailable_Time.get(j))) {
+          unavailable_Time.remove(j);
+        }
+      }
+    }
+
+    //Calculate Free Times
+    int start = TimeRange.START_OF_DAY;
+    int end = TimeRange.END_OF_DAY;
+
+    Collections.sort(unavailable_Time, TimeRange.ORDER_BY_START);
+
+    for (TimeRange time : unavailable_Time) {
+      if (time.start() - start >= request.getDuration()) {
+        available_Time.add(TimeRange.fromStartEnd(start, time.start(), false));
```

Comment:
Here and elsewhere: we should document the boolean argument to make it easier for the code reader to figure out what it is for. 
Here is the way I do it:
`TimeRange.fromStartEnd(start, time.start(), /* inclusive= */ false);`

Reply:
Ya you're right, it would be extra work for someone who doesn't know what that function does